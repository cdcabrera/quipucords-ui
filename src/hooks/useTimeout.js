import { useEffect, useRef, useState } from 'react';
import { helpers } from '../common';

/**
 * setTimeout hook
 *
 * @param {Function} callback
 * @param {number} pollInterval
 * @returns {Function}
 */
const useTimeout = (callback, pollInterval = 0) => {
  const timer = useRef();
  const [update, setUpdate] = useState();
  const result = callback();

  useEffect(() => {
    if (result !== false) {
      timer.current = window.setTimeout(() => setUpdate(helpers.getCurrentDate().getTime()), pollInterval);
    }

    return () => {
      window.clearTimeout(timer.current);
    };
  }, [pollInterval, update, result]);

  return { update, cancel: () => window.clearTimeout(timer.current) };
};

export { useTimeout as default, useTimeout };
