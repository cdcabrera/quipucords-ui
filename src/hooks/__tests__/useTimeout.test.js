import { useTimeout } from '../useTimeout';

describe('useTimeout', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  it('should apply a hook for useTimeout', async () => {
    const mockCallback = jest.fn();
    const mockSetTimeout = jest.spyOn(global, 'setTimeout');
    const { result } = await mountHook(() => useTimeout(mockCallback));

    expect(mockSetTimeout).toHaveBeenCalledTimes(2);
    expect(mockCallback).toHaveBeenCalledTimes(1);
    expect(result).toMatchSnapshot('timeout');
  });
});
