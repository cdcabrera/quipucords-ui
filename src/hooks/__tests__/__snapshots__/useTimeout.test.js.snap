// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTimeout should apply a hook for useTimeout: timeout 1`] = `
Object {
  "cancel": [Function],
  "update": undefined,
}
`;
